main.py

```diff
@@ -102,21 +103,30 @@ def get_configuration_with_overrides(args):
     )
 
 
-def setup_modbus_client(configuration: Configuration) -> ModbusClient:
+def setup_error_handler(
+    configuration: Configuration, mqtt_client: mqtt.Client
+) -> ErrorHandler:
+    return ErrorHandler.from_config(configuration, mqtt_client)
```

Comment:
What do we need a factory method? Why not using the constructor directly instead?

Reply:
The challenge with initialising the ErrorHandler is the variable number of required config settings: either you don't want to publish error messages (in which case `pub_errors` is false and you don't need to provide any other args to init), or you do (in which case you need to init with all the mqtt config settings and the paho client object).

The way I have it currently is that if you are not publishing messages, you use the base `ErrorHandler` class; if you are, you use a subclass `MQTTErrorHandler` which requires all those things in its init method. The factory method determines which class of object you get back, and both init methods require the correct number of args for their mode of operation. Happy to hear suggestions if that's not the best way :)

Reply:
Let's move with this.
If you want to merge the pr as is I can approve it after you update the pr title.

Otherwise, we are missing a demonstration of this working.

It's probably a good time for you to get familiar with the pymodbus server online modification.
We discussed previously you can initiate it in REPL mode and change its behaviour or you can do on the headless start by calling the http endpoint. The latter is good for end to end automation. The former is good for manual t testing.

Can you please ensure to provide evidence of this working as expected?

Reply:
I'm not sure that this is ready to merge yet - I think recent changes have introduced an issue, probably related to having the reader and writer classes connecting to the broker independently. After I publish an error message, the reader seems to lose its topic subscription. Still looking into it.

Reply:
OK that's sorted - the issue was that I was creating one `mqtt_client` and passing it to both reader and writer, which meant the callbacks overwrote each other and messed things up. It's back to working as expected now with two separate client objects.

Reply:
I guess that says we were missing some end to end test then. Happy to add it as a follow up but we should let the pipeline be the one that shows those errors if could happen

Reply:
yes good point - I'll modify the e2e test in #39 to ensure that we re-test receiving commands after publishing an error